File: JAVA/CombinationSumII.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
	ArrayList<ArrayList<Integer>> ret = new ArrayList<ArrayList<Integer>>();
    public ArrayList<ArrayList<Integer>> combinationSum2(int[] num, int target) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ret.clear();
        ArrayList<Integer> path = new ArrayList<Integer>();
        Arrays.sort(num);
        solve(num, 0, num.length, target, path);
        return ret;
    }
	private void solve(int[] num, int index, int len, int target,
			ArrayList<Integer> path) {
		if(target == 0)
		{
			ret.add(new ArrayList<Integer>(path));
			return;
		}
		if(target < 0)
			return;
		for(int i = index; i < len; i++)
		{
			if(i > index && num[i] == num[i-1])
				continue;
			path.add(num[i]);
			solve(num, i+1, len, target - num[i], path);
			path.remove(path.size()-1);
		}
		
	}
}
public class CombinationSumII {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub

	}

}

File: C++/CombinationSumII.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
	vector<vector<int> > ret;
	void dfs(int index, vector<int>& num, int target, vector<int>& path)
	{
		if(target < 0)
			return;
		if(target == 0)
		{
			ret.push_back(vector<int>(path));
			return;
		}
		for(int i = index; i < num.size(); i++)
		{
			if(i != index && num[i] == num[i-1])
				continue;
			path.push_back(num[i]);
			dfs(i+1, num, target - num[i], path);
			path.pop_back();
		}
	}
    vector<vector<int> > combinationSum2(vector<int> &num, int target) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		ret.clear();
		vector<int> path;
		sort(num.begin(), num.end());
		dfs(0, num, target, path);
		return ret;
    }
};